python/python/lance/arrow.py

```diff
@@ -11,14 +11,347 @@
+                raise ValueError("URI {} is not a valid URI".format(uri))
+
+        return cls.from_storage(ImageURIType(), uris)
+
+    def read_uris(self):
+        """
+        Read the images from the URIs into memory and return an EncodedImageArray
+
+        Returns
+        -------
+        EncodedImageArray
+            Array of encoded images
+
+        Examples
+        --------
+        >>> import os
+        >>> uris = [os.path.join(os.path.dirname(__file__), "../tests/images/1.png")]
+        >>> uri_array = ImageURIArray.from_uris(uris)
+        >>> uri_array.read_uris()
+        <lance.arrow.EncodedImageArray object at 0x...>
+        [...]
+        """
+        from pyarrow import fs
+
+        images = []
+        for uri in self.storage:
+            filesystem, path = fs.FileSystem.from_uri(uri.as_py())
+            with filesystem.open_input_stream(path) as f:
+                images.append(f.read())
+
+        return EncodedImageArray.from_storage(
+            EncodedImageType(), pa.array(images, type=pa.binary())
+        )
+
+
+class EncodedImageArray(pa.ExtensionArray):
+    def __repr__(self):
+        return "<lance.arrow.EncodedImageArray object at 0x%016x>\n%s" % (
+            id(self),
+            repr(self.to_pylist()),
+        )
+
+    def image_to_tensor(self, decoder=None):
+        """
+        Decode encoded images and return a FixedShapeImageTensorArray
+
+        Parameters
+        ----------
+        decoder : Callable[pa.binary()], optional
+            A function that takes a pa.binary() and returns a numpy.ndarray
+            or pa.fixed_size_tensor. If not provided, will attempt to use
+            tensorflow or pillow decoder.
+
+        Returns
+        -------
+        FixedShapeImageTensorArray
+            Array of images as tensors
+
+        Examples
+        --------
+        >>> import os
+        >>> uris = [os.path.join(os.path.dirname(__file__), "../tests/images/1.png")]
+        >>> encoded_image_array = ImageURIArray.from_uris(uris).read_uris()
+        >>> encoded_image_array.image_to_tensor()
+        <lance.arrow.FixedShapeImageTensorArray object at 0x...>
+        [[42, 42, 42, 255]]
+        """
+        import numpy as np
+
+        if not decoder:
+
+            def pillow_decoder(x):
+                import io
+
+                from PIL import Image
+
+                return np.stack(
+                    Image.open(io.BytesIO(img.as_py())) for img in self.storage
+                )
+
+            def tensorflow_decoder(x):
+                import tensorflow as tf
+
+                return np.stack(tf.io.decode_image(img.as_py()) for img in self.storage)
+
+            decoders = [
+                ("tensorflow", tensorflow_decoder),
+                ("PIL", pillow_decoder),
+            ]
+            for libname, decoder_function in decoders:
+                try:
+                    __import__(libname)
```

Comment:
what does `__import__` do that's different from `import` keyword?

Reply:
Well this wouldn't work:
```python
libname = "tensorflow"
import libname
```

We could do:
```python
try:
  import tensorflow
  [...]
except:
  try:
    import PIL
    [...]
...
```
But that's less readable.

Reply:
Ah I see. This works fine, just had never seen that before.

Reply:
I'm starting to wonder about my python dialect now after this and the for/else comment :D